keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -55,62 +58,65 @@ class ServerGroupHandler(
   private val publisher: ApplicationEventPublisher,
   override val objectMapper: ObjectMapper,
   override val normalizers: List<ResourceNormalizer<*>>
-) : ResolvableResourceHandler<ServerGroupSpec, ServerGroup> {
+) : ResolvableResourceHandler<ClusterSpec, Set<ServerGroup>> {
 
   override val log: Logger by lazy { LoggerFactory.getLogger(javaClass) }
 
   override val apiVersion = SPINNAKER_API_V1.subApi("ec2")
   override val supportedKind = ResourceKind(
     group = apiVersion.group,
-    singular = "server-group",
-    plural = "server-groups"
-  ) to ServerGroupSpec::class.java
+    singular = "cluster",
+    plural = "clusters"
+  ) to ClusterSpec::class.java
 
-  override suspend fun desired(resource: Resource<ServerGroupSpec>): ServerGroup =
+  override suspend fun desired(resource: Resource<ClusterSpec>): Set<ServerGroup> =
     with(resource.spec) {
-      val (imageId, appVersion) = imageResolver.resolveImageId(resource)
-      ServerGroup(
-        moniker = moniker,
-        location = location,
-        launchConfiguration = launchConfiguration.generateLaunchConfiguration(imageId, appVersion),
-        capacity = capacity,
-        dependencies = dependencies,
-        health = health,
-        scaling = scaling,
-        tags = tags
-      )
+      val resolvedImages = imageResolver.resolveImageId(resource)
+      resolve(resolvedImages)
     }
 
-  override suspend fun current(resource: Resource<ServerGroupSpec>): ServerGroup? =
-    cloudDriverService.getServerGroup(resource)
+  override suspend fun current(resource: Resource<ClusterSpec>): Set<ServerGroup> =
+    cloudDriverService.getServerGroups(resource)
 
   override suspend fun upsert(
-    resource: Resource<ServerGroupSpec>,
-    resourceDiff: ResourceDiff<ServerGroup>
-  ): List<Task> {
-    val spec = resourceDiff.desired
-    val job = when {
-      resourceDiff.isCapacityOnly() -> spec.resizeServerGroupJob(resource.serviceAccount)
-      else -> spec.createServerGroupJob(resource.serviceAccount)
-    }
+    resource: Resource<ClusterSpec>,
+    resourceDiff: ResourceDiff<Set<ServerGroup>>
+  ): List<Task> =
+    resourceDiff
+      .toIndividualDiffs()
```

Comment:
I thought it's easiest to understand what's going on if we handle diffs for regional server groups individually